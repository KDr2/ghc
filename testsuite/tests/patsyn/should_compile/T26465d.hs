
{-# LANGUAGE PatternSynonyms, ViewPatterns #-}

{-# LANGUAGE UnboxedSums, UnboxedTuples, MagicHash #-}

module T26465d where

-- Should-fail variant of T26465c (but with -fdefer-type-errors)

import Data.Kind
  ( Constraint )
import GHC.Exts
  ( TYPE )

type HasP :: forall r. TYPE r -> Constraint
class HasP a where
  getP :: a -> (# (# #) | (# #) #)
  mk :: (# #) -> a

g = getP
m = mk

-- NB: deliberately use no arguments to make this test harder (so that we run
-- into the 'need_dummy_arg' logic of 'GHC.Tc.TyCl.PatSyn.mkPatSynBuilder').
pattern P1 <- ( g -> (# | (# #) #) )
  where P1 = m (# #)

test P1 = P1
