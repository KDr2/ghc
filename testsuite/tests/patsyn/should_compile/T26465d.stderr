T26465d.hs:20:5: warning: [GHC-39999] [-Wdeferred-type-errors (in -Wdefault)]
    • No instance for ‘HasP a0’ arising from a use of ‘getP’
    • In the expression: getP
      In an equation for ‘g’: g = getP

T26465d.hs:21:5: warning: [GHC-39999] [-Wdeferred-type-errors (in -Wdefault)]
    • No instance for ‘HasP a0’ arising from a use of ‘mk’
    • In the expression: mk
      In an equation for ‘m’: m = mk

